File: Sort.py
# Author: Jingze Dai
# Email Address: daij24@mcmaster.ca or david1147062956@163.com
# Github: https://github.com/daijingz
# Linkedin: https://www.linkedin.com/in/jingze-dai/
# Description: Basic Sort
# This file includes programs with 4 different kinds of sorting algorithm
# All of those target inputs are lists consist of unrepeated integers and floats
# Sort output is ascending-integer list
import random


class InappropriateInput(Exception):
    """! Exception when inputs are inappropriate"""
    pass


class OtherErrors(Exception):
    """! Exception for all other problems"""
    pass


def checkAvailable(inputList):
    """! Function Check whether it is correct and appropriate"""
    try:
        i = 0
        while i < len(inputList):
            if not isinstance(inputList[i], (float, int)):
                return False
            i += 1

        for j in inputList:
            if inputList.count(j) > 1:
                return False
            elif j == 0:
                return False

        return True
    except:
        raise OtherErrors("Other errors founded.")


def SelectionSort(inputList):
    """! Sort select the minimum and make it into the new array again and again"""
    try:
        if not checkAvailable(inputList):
            raise InappropriateInput()

        output = []
        loopList = inputList
        while len(loopList) > 0:
            output += [min(loopList)]
            loopList.remove(min(loopList))
        return output
    except:
        raise OtherErrors("Other errors founded.")


def InsertionSort(inputList):
    """! Sort select elements from left to right and insert them sequentially"""
    try:
        if not checkAvailable(inputList):
            raise InappropriateInput()

        output = []
        for i in inputList:
            output = Insertion(output, i)
        return output
    except:
        raise OtherErrors("Other errors founded.")


def Insertion(inputList, elem):
    """! Find the right place for an element in a sorted list"""
    try:
        if not checkAvailable(inputList):
            raise InappropriateInput()

        if len(inputList) == 0:
            return [elem]
        elif elem > max(inputList):
            return inputList + [elem]
        elif elem < min(inputList):
            return [elem] + inputList
        else:
            output = inputList
            i = 0
            while i < len(inputList):
                if elem < output[i]:
                    output = output[:i] + [elem] + output[i:]
                    return output
                else:
                    i += 1
    except:
        raise OtherErrors("Other errors founded.")


def MergeSort(inputList):
    """! Divide list into sub-lists and sort them, then combine their results together"""
    try:
        if not checkAvailable(inputList):
            raise InappropriateInput()

        if len(inputList) > 1:
            subLeft = inputList[:len(inputList) // 2]
            subRight = inputList[len(inputList) // 2:]

            MergeSort(subLeft)
            MergeSort(subRight)

            i = 0
            j = 0
            k = 0

            while i < len(subLeft) and j < len(subRight):
                if subLeft[i] < subRight[j]:
                    inputList[k] = subLeft[i]
                    i += 1
                elif subLeft[i] > subRight[j]:
                    inputList[k] = subRight[j]
                    j += 1
                k += 1

            while i < len(subLeft):
                inputList[k] = subLeft[i]
                i += 1
                k += 1

            while j < len(subRight):
                inputList[k] = subRight[j]
                j += 1
                k += 1

            output = inputList
            return output
        else:
            return inputList
    except:
        raise OtherErrors("Other errors founded.")


def ShellSort(inputPart):
    """! Variation of Insertion Sort"""
    try:
        if not checkAvailable(inputPart):
            raise InappropriateInput()

        inputList = inputPart
        medium = len(inputList) // 2
        while medium > 0:
            for i in range(medium, len(inputList)):
                temp = inputList[i]
                j = i
                while j >= medium and inputList[j - medium] > temp:
                    inputList[j] = inputList[j - medium]
                    j -= medium
                inputList[j] = temp
            medium //= 2
        return inputList
    except:
        raise OtherErrors("Other errors founded.")


def QuickSort(inputList):
    """! Find a comparable index (not max or min) and then compare others"""
    try:
        if not checkAvailable(inputList):
            raise InappropriateInput()

        leftPart = []
        rightPart = []
        randomNum = inputList[random.randint(0, len(inputList) - 1)]

        while randomNum == max(inputList) or randomNum == min(inputList):
            randomNum = inputList[random.randint(0, len(inputList) - 1)]

        for i in inputList:
            if i < randomNum:
                leftPart += [i]
            elif i > randomNum:
                rightPart += [i]

        leftPart.sort()
        rightPart.sort()
        return leftPart + [randomNum] + rightPart
    except:
        raise OtherErrors("Other errors founded.")

File: Sort.java
// Array Numbers must be unrepeated

public class Sort{
    int[] Selection_Sort(int[] arr) {
        int n = arr.length;
        for (int i = 0; i < n-1; i++){
            int min_idx = i;
            for (int j = i+1; j < n; j++)
                if (arr[j] < arr[min_idx])
                    min_idx = j;

            int temp = arr[min_idx];
            arr[min_idx] = arr[i];
            arr[i] = temp;
        }
        return arr;
    }

    int[] Insertion_Sort(int[] arr){
        int n = arr.length;
        for (int i = 1; i < n; ++i) {
            int key = arr[i];
            int j = i - 1;

            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j = j - 1;
            }
            arr[j + 1] = key;
        }
        return arr;
    }

    void merge(int arr[], int l, int m, int r) {
        int n1 = m - l + 1;
        int n2 = r - m;

        int L[] = new int[n1];
        int R[] = new int[n2];

        for (int i = 0; i < n1; ++i) {
            L[i] = arr[l + i];
        }
        for (int j = 0; j < n2; ++j) {
            R[j] = arr[m + 1 + j];
        }

        int i = 0, j = 0;
        int k = l;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            }
            else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    int[] Merge_Sort(int arr[], int l, int r) {
        if (l < r) {
            int m =l+ (r-l)/2;
            Merge_Sort(arr, l, m);
            Merge_Sort(arr,m + 1, r);
            merge(arr, l, m, r);
        }
        return arr;
    }

    int[] Shell_Sort(int[] arr) {
        int n = arr.length;
        for (int gap = n/2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i += 1) {
                int temp = arr[i];
                int j;
                for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
                    arr[j] = arr[j - gap];
                arr[j] = temp;
            }
        }
        return arr;
    }

    int SelectMIN(int[] input){
        int output = 100000;
        for (int v : input) {
            if (v < output) {
                output = v;
            }
        }
        return output;
    }

    int IndexMIN(int[] input){
        int num = 100000;
        int output = 0;
        for (int i = 0; i < input.length; i++) {
            if (input[i] < num) {
                num = input[i];
                output = i;
            }
        }
        return output;
    }

    double SelectMAX(double[] input){
        double output = -100000.0;
        for (double v : input) {
            if (v > output) {
                output = v;
            }
        }
        return output;
    }

    int IndexMAX(double[] input){
        double num = -100000.0;
        int output = 0;
        for (int i = 0; i < input.length; i++) {
            if (input[i] > num) {
                num = input[i];
                output = i;
            }
        }
        return output;
    }
}
